File: sol/380.py
from random import choice

class RandomizedSet:

    def __init__(self):
        self.index: dict[int, int] = dict()
        self.items: list[int] = list()

    def insert(self, val: int) -> bool:
        if val in self.index:
            return False
        self.index[val] = len(self.items)
        self.items.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.index:
            return False
        last = self.items.pop()
        if last != val:
            self.index[last] = self.index[val]
            self.items[self.index[last]] = last
        del self.index[val]
        return True

    def getRandom(self) -> int:
        return choice(self.items)

File: sol/380.cpp
#include <unordered_map>
#include <random>

struct RandomizedSet {
    vector<int> items;
    unordered_map<int, int> index;
    mt19937 g;

    RandomizedSet(): g(std::random_device{}()) {}
    
    bool insert(int val) {
        if (this->index.contains(val)){
            return false;
        }
        this->index[val] = this->items.size();
        this->items.push_back(val);
        return true;
    }
    
    bool remove(int val) {
        if (!this->index.contains(val)){
            return false;
        }
        int last = this->items.back();
        this->items.pop_back();
        if (last != val){
            int k = this->index[val];
            this->index[last] = k;
            this->items[k] = last;
        }
        this->index.erase(val);
        return true;
    }
    
    int getRandom(void) {
        return this->items[this->g() % this->items.size()];
    }
};

